databroker/src/grpc/kuksa_val_v2/val.rs

```diff
@@ -39,22 +39,96 @@ const MAX_REQUEST_PATH_LENGTH: usize = 1000;
 impl proto::val_server::Val for broker::DataBroker {
     async fn get_value(
         &self,
-        _request: tonic::Request<proto::GetValueRequest>,
+        request: tonic::Request<proto::GetValueRequest>,
     ) -> Result<tonic::Response<proto::GetValueResponse>, tonic::Status> {
-        Err(tonic::Status::new(
-            tonic::Code::Unimplemented,
-            "Unimplemented",
-        ))
+        debug!(?request);
+        let permissions = match request.extensions().get::<Permissions>() {
+            Some(permissions) => {
+                debug!(?permissions);
+                permissions.clone()
+            }
+            None => return Err(tonic::Status::unauthenticated("Unauthenticated")),
+        };
+
+        let broker = self.authorized_access(&permissions);
+
+        let request = request.into_inner();
+
+        let signal_id = match get_signal(request.signal_id, &broker).await {
+            Ok(signal_id) => signal_id,
+            Err(err) => return Err(err),
+        };
+
+        let datapoint = match broker.get_datapoint(signal_id).await {
+            Ok(datapoint) => datapoint,
+            Err(ReadError::NotFound) => return Err(tonic::Status::not_found("Path not found")),
+            Err(ReadError::PermissionDenied) => {
+                return Err(tonic::Status::permission_denied("Permission denied"))
+            }
+            Err(ReadError::PermissionExpired) => {
+                return Err(tonic::Status::unauthenticated("Unauthenticated"))
+            }
+        };
+
+        Ok(tonic::Response::new(proto::GetValueResponse {
+            data_point: datapoint.into(),
+        }))
     }
 
     async fn get_values(
         &self,
-        _request: tonic::Request<proto::GetValuesRequest>,
+        request: tonic::Request<proto::GetValuesRequest>,
     ) -> Result<tonic::Response<proto::GetValuesResponse>, tonic::Status> {
-        Err(tonic::Status::new(
-            tonic::Code::Unimplemented,
-            "Unimplemented",
-        ))
+        debug!(?request);
+        let permissions = match request.extensions().get::<Permissions>() {
+            Some(permissions) => {
+                debug!(?permissions);
+                permissions.clone()
+            }
+            None => return Err(tonic::Status::unauthenticated("Unauthenticated")),
+        };
+
+        let broker = self.authorized_access(&permissions);
+
+        let requested = request.into_inner().signal_ids;
+        let mut response_datapoints = Vec::new();
```

Comment:
shouldn't we have a map here instead. Since it's async the order is not guaranteed to be the same as requested. Maybe a general discussion point as well @rafaeling @argerus

Reply:
Why would the order be affected by whether it's async? Or exactly what do you mean by "Since it's async"?

Reply:
In the Proto file we say that the result shall be returned in the same order as requested and here it gets the values one by one so should work. I have some test cases which at least check that it work for two signals (Which however does not prove anything)

Reply:
Yeah you're right I've missed the await call all good will resolve this conversation :)